sql: enable byte arrays to be evaluated from a constant

otan opened the pull request:
Resolves #71444

Release note (sql change): Previously, comparing against `bytea[]`
without a cast (e.g. `SELECT * FROM t WHERE byteaarrcol = '{}'`) would
result in an ambiguous error. This has now been resolved.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71501)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice, just a random question!

otan committed:
sql: enable byte arrays to be evaluated from a constant

Release note (sql change): Previously, comparing against `bytea[]`
without a cast (e.g. `SELECT * FROM t WHERE byteaarrcol = '{}'`) would
result in an ambiguous error. This has now been resolved.
